### ISSUE TITLE
Support for managing revoked certs

### ISSUE THREAD

**Issue by stensonb** [kubernetes/kubernetes#18982]
Authenticating users via x509 certs is important, but the project seems to be missing a mechanism to revoke certs (without throwing the entire chain away and regenerating ALL certs for all users).

It would be great to be able to declare which certs are invalid, and have the kube-apiserver, kubelets, and all other cert-dependent services deny service for requests with the now-invalid cert.

https://en.wikipedia.org/wiki/OCSP_stapling appears to be one way to solve this.

FYI - this is the same idea as the issue with CoreOS: https://github.com/coreos/etcd/issues/4034

**Comment by roberthbailey** [comment]
@stensonb That is a good observation. 

/cc @ncdc @stephenR

**Comment by ncdc** [comment]
cc @liggitt

**Comment by jimmycuadra** [comment]
Somewhat related, a better story around managing the lifecycle of TLS certs/keys in general. See https://github.com/kubernetes/kubernetes/issues/25379.

**Comment by rikatz** [comment]
+1

**Comment by tmegow** [comment]
+1

**Comment by littleNMm** [comment]
+1

**Comment by kaazoo** [comment]
+1

**Comment by mattmoyer** [comment]
As a start, we could add a new flag `--client-ca-revocation-list-file=/path/to/crl.pem` to point at a local CRL file. This would allow for manual revocation of certs which would meet many use cases and matches the solution that ended up in etcd (https://github.com/coreos/etcd/pull/8124).

To revoke a certificate, a user would:

1. Sign a CRL using the CA (or a dedicated CRL issuer).
1. Get the CRL file onto the control plane node where the apiserver is running.
1. Update the apiserver flag to point at the file and restart the apiserver process.

These steps could be automated in `kubeadm` and other tools.

This would cover, for example, the case where my installer generated a long-lived `kubernetes-admin` certificate that I want to replace with individual admin user identities.

**Comment by mikedanese** [comment]
* Who keeps track of serial numbers? The [local signer](https://github.com/cloudflare/cfssl/blob/275fb308ac705bf5631d23b941f7b56dc436e39d/signer/local/local.go#L295-L296) doesn't. How do you reasonably construct CRLs with the kubeadm setup?
* How long is the CRL valid for? What regenerates the CRL when it is close to expiring? Kubeadm is a main, not a daemon. Do we fail open or closed when CRLs become invalid?
* What about kubelet handlers? How do we distribute CRLs to nodes?

**Comment by mikedanese** [comment]
I would prefer OCSP stapling with a validation authority that had some plugable policy engine, e.g. OPA.

**Comment by simo5** [comment]
OCSP works great for serving certs, not so great for client certs though ...

**Comment by simo5** [comment]
Smart cards are often not smart enough :-)
Besides that I am not a aware of an RFC that defines how to do OCSP stapling for client certs, both RFC6066 and RFC6961 only describe how to ask stapled responses from a server, I see no mechanism to ask for a stapled response from a client.

**Comment by simo5** [comment]
Unless you meant that servers will do OCSP stapling, while client certs are validated via OCSP, that is possible I guess, but error prone as it relies on the OCSP responder to be highly available again.

In general it seems like the best scenario is to use OCSP stapling for server certs and CRLs for client certs, as that keeps all the work on the server side, and can all be done ahead of time.

**Comment by droberin** [comment]
I think thi is a very important matter, yet it is almost a 3 years old issue at GitHub and it doesn't seem to be addressed/assign. I'm just curious about if it is not really that important or it is too big to take care of.
Thanks and sorry for the noise. :)

**Comment by joewilliams** [comment]
As with the other commenters I am also very interested in the status on this issue.

**Comment by sbinev** [comment]
Ability to revoke (at least apiserver) certificates is paramount!

**Comment by kron4eg** [comment]
Couple of examples I can think of immediately:
* https://golang.org/pkg/crypto/x509/#ParseCRL
* https://github.com/cloudflare/cfssl/blob/master/crl/crl.go

**Comment by sftim** [comment]
A few thoughts.

1. OSCP stapling could work via periodic updates to CertificateSigningRequest objects. That way, parts of the cluster can use long-lived certificates and also let clients benefit from stapled OCSP responses.
   For example, stapling an OCSP response that says that the certificate is OK for the next 300 seconds.
   OCSP stapling is useful for clients outside the cluster (eg `kubectl`), if it's supported.
2. A new CertificateRevocationRequest object could parallel CertificateSigningRequests. For example, a evocation-approval controller might accept any CertificateRevocationRequest that carries a signature made by the certificate being revoked.
3. If CertificateRevocationRequest objects (or similar) are available via the API server, then the API server can reject client connections that are authenticated with any revoked certificate.
4. A controller count track revocations and publish a revocation list (CRL) that components such as Kubelet can consume.
5. Another controller can obtain signed, timestamped OCSP responses for certificates such as the API server's main certificate, and publish those via the Kubernetes API so that other components can staple those to their TLS sessions.

Overall it sounds possible and even worthwhile, but also quite a lot of work.

**Comment by oponder** [comment]
Would it be super hacky to have a flag on the api server that allows us to reject certs created BEFORE a certain date?

That would be a way to quickly revoke all certs and start over when you have a feeling something went wrong.

**Comment by ep4eg** [comment]
> Would it be super hacky to have a flag on the api server that allows us to reject certs created BEFORE a certain date?
> 
> That would be a way to quickly "revoke" all certs and start over when you have a feeling something went wrong.

So that means any time I want to revoke a cert I have to revoke all of them and sign another bunch? Doesn't sound sane for me.

**Comment by oponder** [comment]
> So that means any time I want to revoke a cert I have to revoke all of them and sign another bunch? Doesn't sound sane for me.

My use case is indeed to revoke all certs and start over, without having to change the CA or revoke an intermediate certificate, since that is a lot more work, or not directly in people's control.

**Comment by chris13524** [comment]
Alternative to revoking certificates, one could implement [Webhook access control](https://kubernetes.io/docs/reference/access-authn-authz/webhook/) in a higher priority than Node or RBAC which would enable immediate denial of access as well as denial of later certificate renewal.

So in affect one can disable authorization but not authentication.

Are there any security concerns with having a valid certificate but no privileges?

**Comment by raesene** [comment]
The is a potential concern with valid certificate and no explicit grants in RBAC, which is that the user will still get any rights assigned to the `system:authenticated` group. Whether that presents a problem, will likely depend on how the cluster operator manages RBAC in their environment, but it's not ideal.

**Comment by enj** [comment]
@deads2k are you planning on a KEP related to this in the v1.23 release?

**Comment by sferich888** [comment]
> @deads2k are you planning on a KEP related to this in the v1.23 release?

v1.23 has come and past and is well on its way to [retirement (targeted for End of Life:2023-02-28](https://kubernetes.io/releases/#release-v1-23)); does anyone have any plans to address this or propose a change that could be reviewed?

**Comment by remram44** [comment]
As far as I can see, this can be done the same way etcd did it, by overriding the TLS Listener with one that checks against a revocation list.

etcd's check: https://github.com/heyitsanthony/etcd/blob/41e26f741b26cc6f3faa39151ef74cfee3b6eace/pkg/transport/listener_tls.go#L62-L74

apiserver listener creation: https://github.com/kubernetes/apiserver/blob/10c70bebf96a41cfc1fb721c80a668d648cfeb0c/pkg/server/secure_serving.go#L250

Being able to pass a revocation list via an apiserver command-line option would already go a long way.

**Comment by XANi** [comment]
Is there any chance to get even basic "Here is my CRL, use my CRL" included ? There appears to be half a dozen PR about exact same issue since 2016 all closed by indecisiveness. 

"Here is  CRL file" and "here is a signal for API server to re-read CRL" at least allows to have a solution for problem existing. 

"Here is CRL file URL, just download it every X minutes or every crl.NextUpdate" would probably solve 99% of the problem as then we can just point it at whatever cert management solution the cluster is using

**Comment by sftim** [comment]
This issue is open; pull requests are welcome.

**Comment by remram44** [comment]
I sent a pull request, would love feedback: #122203

**Comment by enj** [comment]
> This issue is open; pull requests are welcome.

This isn't an issue that one can just make a PR to fix.  It requires agreement on the path forward.  The code itself was never the limiting factor.

**Comment by raesene** [comment]
@enj do you know is there any documentation around previous discussions on options for looking at this that have been considered and decided against? I could see a challenge here for people coming at this fresh, that they won't have the perspective of the past 8 years to look back on.

I can recall some discussions in SIG-Auth slack, but I'm not sure if there's other docs.

**Comment by tzakrajs** [comment]
This is an important feature, it’s too bad the maintainers aren’t interested in continuing the discussion so that we can actually get a PR landed after almost 10 years of waiting.

**Comment by XANi** [comment]
Yeah, we were making bot to ship the certs out to our cluster user then realized "wait, if someone gets their device stolen or hacked into there is literally no way to revoke their certs".

At this point using cert auth is worse than plain passwords. Hell, with just normal login it could be driven off corporate LDAP...

**Comment by Doctor-love** [comment]
Meanwhile I've hacked together a partial work-around for lacking revocation checks: https://github.com/menacit/k8s_crlish_authorizer/

Hope this is useful/could serve as inspiration to someone!

**Comment by remram44** [comment]
@Doctor-love: oh that looks good! I was not aware of this possibility. Will your tool work together with other authentication methods, e.g. can I have this set up with OIDC also enabled?

**Comment by Doctor-love** [comment]
> [@Doctor-love](https://github.com/Doctor-love): oh that looks good! I was not aware of this possibility. Will your tool work together with other authentication methods, e.g. can I have this set up with OIDC also enabled?

Not 100% sure, but should work for OIDC as well, [as long as the IdP sets the "jti" claim](https://github.com/kubernetes/kubernetes/blob/a552570d2c0d5a2589c402695c74e1a85deed617/staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go#L962-L971). Works for time-limited service account tokens as well.

**Comment by enj** [comment]
> Meanwhile I've hacked together a partial work-around for lacking revocation checks: https://github.com/menacit/k8s_crlish_authorizer/

@Doctor-love note that any identity with the `system:masters` group will always be allowed to perform all actions before any authorization webhook is called.

**Comment by santiago-mooser** [comment]
@mikedanese is this something that is planned in the future or not at all?

**Comment by liggitt** [comment]
There are not plans to add this.

**Comment by lentzi90** [comment]
I just wanted to highlight that the issue description mentions OCSP stapling (https://en.wikipedia.org/wiki/OCSP_stapling) but the current discussion seem to be only about CRL. Has there been any investigations or discussions in this direction?

As I see it, OCSP stapling would address the concerns of 1 (availability, complexity and runtime cost) by pushing the cost on the clients instead. Server side verification would be straight forward. Any thoughts about this?

**Comment by remram44** [comment]
Having to run an OCSP server and handle many requests to it just to handle the couple of certificates that I will ever revoke... We lost the plot.

**Comment by XANi** [comment]
Exactly. CRL works just fine in small controlled environments and requires zero changes on client side or spinning up new servers and distributing single CRL file to a bunch of k8s servers is trivial for any org with any half decent configuration management setup

**Comment by lentzi90** [comment]
> Having to run an OCSP server and handle many requests to it just to handle the couple of certificates that I will ever revoke...

I can understand this perspective as well. For an operator, the OCSP alternative probably means more work. However, I think the main concern here is around Kubernetes as a project. I would be interested to know if someone has looked into the feasibility of it. Perhaps it would be easier, even though it might mean more operational overhead outside of Kubernetes.

> Exactly. CRL works just fine in small controlled environments and requires zero changes on client side or spinning up new servers and distributing single CRL file to a bunch of k8s servers is trivial for any org with any half decent configuration management setup

What about larger, complex and dynamic environments? Bare metal, air-gapped deployments with external CA? I am not saying CRL is a no-go, just that "small controlled environments" may not be where this is needed.

**Comment by StephanX** [comment]
>What about larger, complex and dynamic environments? 

Make a ball park guess at how many revoked certificates might be in a large, complex, dynamic environment?  Dozens? Maybe even a couple hundred?  Like, say, less than a kilobyte worth of data?  A _trivial_ amount of data to reference and propagate?

That's the frustration with this _nine_ year old issue; solving the problem seems ridiculously easy, but the scope around how to approach it has ballooned to a point where it doesn't seem like it will ever be seriously addressed.

**Comment by raesene** [comment]
One interesting point here is that the Kubernetes project already has a pattern for effective revocation of credentials, which is having another object which is referenced from the credential and when the referenced object is removed, the credential is invalidated. This pattern works in Service account tokens and pod certificates.

One option might have been extending that pattern to have a referenced object for all issued certificated (e.g. a certificate object) and just allow admins to delete the object for certs that are to be revoked. This would have the side benefit of allowing for credential review/auditing, which would be nice.

**Comment by lentzi90** [comment]
I looked around regarding OCSP and it seems it has very limited support in go. Found these two issues related to it:
- https://github.com/golang/go/issues/22274
- https://github.com/golang/go/issues/40521

So that is out then for the foreseeable future I guess.

**Comment by remram44** [comment]
> One interesting point here is that the Kubernetes project already has a pattern for effective revocation of credentials, which is having another object which is referenced from the credential and when the referenced object is removed, the credential is invalidated. This pattern works in Service account tokens and pod certificates.

This is interesting! Indeed, if I could just delete the Node to lock out a kubelet, that would solve the entire problem for me. Unfortunately what actually happens now is the kubelet holding a valid certificate can recreate its Node and re-join.

**Comment by XANi** [comment]
@lentzi90 
> What about larger, complex and dynamic environments? Bare metal, air-gapped deployments with external CA? I am not saying CRL is a no-go, just that "small controlled environments" may not be where this is needed.

Even at scale of 10k developers (let's assume they all use same cluster and you use good practice of short lived certs) you'd still have *maybe* 1k certs on the list at any given time, that's few tens of kb and an amortized O(1) lookup in a hash

**Comment by lentzi90** [comment]
> Broadly, I see ~3 possible approaches:
> 
> 1. Implement CRL lookups based on content of presented certificates. When this was discussed in the past, the complexity, availability risk, and runtime cost of letting presented certificates drive remote lookups server-side was considered a non-starter
>
> 2. Design an administrator-configured way of revoking / blocking use of certificates (static CRL, or bespoke list of disallowed certificate serial numbers or subjects or fingerprints, etc). This would be a decent amount of effort that would only enable a very narrow capability. I wouldn't be in favor of this approach.
>
> 3. A general way for admins to validate the output of the authentication layer. We already have this for JWT authentication (search for `userValidationRules` examples in the [documentation](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#using-authentication-configuration)), and there has been discussion ([2025-09-10 sig-auth](https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM/edit?tab=t.0#bookmark=id.t6tymtxhw4ra)) of expanding this capability to be able to see / verify authentications from any source, including x509 authentication. Since x509 authentication makes information about the client certificate visible in the authenticated user (as an `authentication.k8s.io/credential-id` field set to `X509SHA256=<SHA256-hash-of-certificate>`), this _could_ be used to revoke particular certificates. This would seem the most promising direction to me, putting effort into a capability that would allow for blocking use of specific certificates along with many other use cases.

I have tried to explore alternative 3 from @liggitt 's comment above, in https://github.com/kubernetes/kubernetes/pull/139812. So far it seems to work well for me and I have gotten encouraging feedback internally also.
It is obviously just exploratory work and hence pushed as a draft. I would be very interested in your thoughts though.
Does the approach look valid? Is this something we could move forward with? Do we need a KEP in that case?
